lib/core/networking/server.dart

```diff
@@ -0,0 +1,156 @@
+import 'dart:async';
+import 'dart:io';
+
+import 'websockets/session.dart';
+import 'websockets/middleware.dart';
+
+class WebServer {
+  /// Awaitable to ensure ready for use
+  late Future<void> ready;
+
+  /// List of middleware that should be installed in each WebSocket session.
+  ///
+  /// Each middleware recieves the calling [Session] instance, message,
+  /// [MiddlewareAction] and a proxy to the next [MiddlewareFunc].
+  List<MiddlewareFunc> wsMiddleware;
+
+  /// Handler for received WebSocket messages
+  Future<dynamic> Function(ServerSession, dynamic)? wsOnMessage;
+
+  /// Handler for encountered WebSocket errors
+  void Function(ServerSession, Object)? wsOnError;
+
+  /// Handler for stream WebSocket closure
+  void Function(ServerSession)? wsOnDone;
+
+  /// Handler for received HTTP requests.
+  ///
+  /// NOTE: Please ensure this calls `close()` on the response
+  void Function(HttpRequest)? httpOnRequest;
+
+  /// If WebSocket upgrade requests should be redirected
+  /// to the HTTP handler if no [wsOnMessage] was specified.
+  bool redirectWebsocket;
+
+  /// Handler for stream WebSocket closure that wraps [wsOnDone]
+  /// but also removes the session from the server.
+  late void Function(ServerSession) wsOnDoneTeardown;
+
+  /// HttpServer instance used to establish initial connection
+  HttpServer? _server;
```

Comment:
No, the getters `address` and `port` are not async, if the user didn't await `await server.ready;` `_server` will be `null`.

Reply:
So it's intendet beaviour that `WebServer` has no port and no address? Awaiting server.ready is mandatory and there's no use in getting `null` as port or address. Throwing an error instead of `null` will give you the benefit of avoiding the usage of uninitialized servers.